File: nodejs/test/e2e/session_config.e2e.test.ts
import { describe, expect, it } from "vitest";
import { writeFile, mkdir } from "fs/promises";
import { join } from "path";
import { approveAll } from "../../src/index.js";
import { createSdkTestContext } from "./harness/sdkTestContext.js";

describe("Session Configuration", async () => {
    const { copilotClient: client, workDir, openAiEndpoint } = await createSdkTestContext();

    it("should use workingDirectory for tool execution", async () => {
        const subDir = join(workDir, "subproject");
        await mkdir(subDir, { recursive: true });
        await writeFile(join(subDir, "marker.txt"), "I am in the subdirectory");

        const session = await client.createSession({
            onPermissionRequest: approveAll,
            workingDirectory: subDir,
        });

        const assistantMessage = await session.sendAndWait({
            prompt: "Read the file marker.txt and tell me what it says",
        });
        expect(assistantMessage?.data.content).toContain("subdirectory");

        await session.disconnect();
    });

    it("should create session with custom provider config", async () => {
        const session = await client.createSession({
            onPermissionRequest: approveAll,
            provider: {
                baseUrl: "https://api.example.com/v1",
                apiKey: "test-key",
            },
        });

        expect(session.sessionId).toMatch(/^[a-f0-9-]+$/);

        try {
            await session.disconnect();
        } catch {
            // disconnect may fail since the provider is fake
        }
    });

    it("should accept blob attachments", async () => {
        // Write the image to disk so the model can view it if it tries
        const pngBase64 =
            "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNk+M9QDwADhgGAWjR9awAAAABJRU5ErkJggg==";
        await writeFile(join(workDir, "pixel.png"), Buffer.from(pngBase64, "base64"));

        const session = await client.createSession({ onPermissionRequest: approveAll });

        await session.sendAndWait({
            prompt: "What color is this pixel? Reply in one word.",
            attachments: [
                {
                    type: "blob",
                    data: pngBase64,
                    mimeType: "image/png",
                    displayName: "pixel.png",
                },
            ],
        });

        await session.disconnect();
    });

    it("should accept message attachments", async () => {
        await writeFile(join(workDir, "attached.txt"), "This file is attached");

        const session = await client.createSession({ onPermissionRequest: approveAll });

        await session.sendAndWait({
            prompt: "Summarize the attached file",
            attachments: [{ type: "file", path: join(workDir, "attached.txt") }],
        });

        await session.disconnect();
    });

    const PNG_1X1 = Buffer.from(
        "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNk+M9QDwADhgGAWjR9awAAAABJRU5ErkJggg==",
        "base64"
    );
    const VIEW_IMAGE_PROMPT =
        "Use the view tool to look at the file test.png and describe what you see";

    function hasImageUrlContent(messages: Array<{ role: string; content: unknown }>): boolean {
        return messages.some(
            (m) =>
                m.role === "user" &&
                Array.isArray(m.content) &&
                m.content.some((p: { type: string }) => p.type === "image_url")
        );
    }

    it("vision disabled then enabled via setModel", async () => {
        await writeFile(join(workDir, "test.png"), PNG_1X1);

        const session = await client.createSession({
            onPermissionRequest: approveAll,
            modelCapabilities: { supports: { vision: false } },
        });

        // Turn 1: vision off — no image_url expected
        await session.sendAndWait({ prompt: VIEW_IMAGE_PROMPT });
        const trafficAfterT1 = await openAiEndpoint.getExchanges();
        const t1Messages = trafficAfterT1.flatMap((e) => e.request.messages ?? []);
        expect(hasImageUrlContent(t1Messages)).toBe(false);

        // Switch vision on (re-specify same model with updated capabilities)
        await session.setModel("claude-sonnet-4.5", {
            modelCapabilities: { supports: { vision: true } },
        });

        // Turn 2: vision on — image_url expected
        await session.sendAndWait({ prompt: VIEW_IMAGE_PROMPT });
        const trafficAfterT2 = await openAiEndpoint.getExchanges();
        // Only check exchanges added after turn 1
        const newExchanges = trafficAfterT2.slice(trafficAfterT1.length);
        const t2Messages = newExchanges.flatMap((e) => e.request.messages ?? []);
        expect(hasImageUrlContent(t2Messages)).toBe(true);

        await session.disconnect();
    });

    it("vision enabled then disabled via setModel", async () => {
        await writeFile(join(workDir, "test.png"), PNG_1X1);

        const session = await client.createSession({
            onPermissionRequest: approveAll,
            modelCapabilities: { supports: { vision: true } },
        });

        // Turn 1: vision on — image_url expected
        await session.sendAndWait({ prompt: VIEW_IMAGE_PROMPT });
        const trafficAfterT1 = await openAiEndpoint.getExchanges();
        const t1Messages = trafficAfterT1.flatMap((e) => e.request.messages ?? []);
        expect(hasImageUrlContent(t1Messages)).toBe(true);

        // Switch vision off
        await session.setModel("claude-sonnet-4.5", {
            modelCapabilities: { supports: { vision: false } },
        });

        // Turn 2: vision off — no image_url expected in new exchanges
        await session.sendAndWait({ prompt: VIEW_IMAGE_PROMPT });
        const trafficAfterT2 = await openAiEndpoint.getExchanges();
        const newExchanges = trafficAfterT2.slice(trafficAfterT1.length);
        const t2Messages = newExchanges.flatMap((e) => e.request.messages ?? []);
        expect(hasImageUrlContent(t2Messages)).toBe(false);

        await session.disconnect();
    });

    const PROVIDER_HEADER_NAME = "x-copilot-sdk-provider-header";
    const CLIENT_NAME = "ts-public-surface-client";

    function createProxyProvider(headerValue: string) {
        return {
            type: "openai" as const,
            baseUrl: openAiEndpoint.url,
            apiKey: "test-provider-key",
            headers: {
                [PROVIDER_HEADER_NAME]: headerValue,
            },
        };
    }

    function getHeaderString(
        headers: Record<string, string | string[] | undefined> | undefined,
        name: string
    ): string | undefined {
        if (!headers) {
            return undefined;
        }
        const matchingKey = Object.keys(headers).find(
            (k) => k.toLowerCase() === name.toLowerCase()
        );
        if (!matchingKey) {
            return undefined;
        }
        const value = headers[matchingKey];
        if (Array.isArray(value)) {
            return value.join(",");
        }
        return value ?? "";
    }

    function getSystemMessage(exchange: {
        request: { messages?: Array<{ role: string; content: unknown }> };
    }): string | undefined {
        const sys = (exchange.request.messages ?? []).find((m) => m.role === "system") as
            | { content: string }
            | undefined;
        return sys?.content;
    }

    function getToolNames(exchange: {
        request: { tools?: Array<{ function: { name: string } }> };
    }): string[] {
        return (exchange.request.tools ?? []).map((t) => t.function.name);
    }

    it("should forward clientName in user-agent", async () => {
        const session = await client.createSession({
            onPermissionRequest: approveAll,
            clientName: CLIENT_NAME,
        });

        await session.sendAndWait({ prompt: "What is 1+1?" });

        const exchanges = await openAiEndpoint.getExchanges();
        expect(exchanges.length).toBeGreaterThan(0);
        const userAgent = getHeaderString(exchanges[0].requestHeaders, "user-agent");
        expect(userAgent).toBeDefined();
        expect(userAgent).toContain(CLIENT_NAME);

        await session.disconnect();
    });

    it("should forward custom provider headers on create", async () => {
        const session = await client.createSession({
            onPermissionRequest: approveAll,
            model: "claude-sonnet-4.5",
            provider: createProxyProvider("create-provider-header"),
        });

        const message = await session.sendAndWait({ prompt: "What is 1+1?" });
        expect(message?.data.content ?? "").toContain("2");

        const exchanges = await openAiEndpoint.getExchanges();
        expect(exchanges.length).toBeGreaterThan(0);
        const auth = getHeaderString(exchanges[0].requestHeaders, "authorization");
        expect(auth).toContain("Bearer test-provider-key");
        const customHeader = getHeaderString(exchanges[0].requestHeaders, PROVIDER_HEADER_NAME);
        expect(customHeader).toContain("create-provider-header");

        await session.disconnect();
    });

    it("should forward custom provider headers on resume", async () => {
        const session1 = await client.createSession({ onPermissionRequest: approveAll });
        const sessionId = session1.sessionId;

        const session2 = await client.resumeSession(sessionId, {
            onPermissionRequest: approveAll,
            model: "claude-sonnet-4.5",
            provider: createProxyProvider("resume-provider-header"),
        });

        const message = await session2.sendAndWait({ prompt: "What is 2+2?" });
        expect(message?.data.content ?? "").toContain("4");

        const exchanges = await openAiEndpoint.getExchanges();
        expect(exchanges.length).toBeGreaterThan(0);
        const lastExchange = exchanges[exchanges.length - 1];
        const auth = getHeaderString(lastExchange.requestHeaders, "authorization");
        expect(auth).toContain("Bearer test-provider-key");
        const customHeader = getHeaderString(lastExchange.requestHeaders, PROVIDER_HEADER_NAME);
        expect(customHeader).toContain("resume-provider-header");

        await session2.disconnect();
    });

    it("should apply workingDirectory on session resume", async () => {
        const subDir = join(workDir, "resume-subproject");
        await mkdir(subDir, { recursive: true });
        await writeFile(join(subDir, "resume-marker.txt"), "I am in the resume working directory");

        const session1 = await client.createSession({ onPermissionRequest: approveAll });
        const sessionId = session1.sessionId;

        const session2 = await client.resumeSession(sessionId, {
            onPermissionRequest: approveAll,
            workingDirectory: subDir,
        });

        const message = await session2.sendAndWait({
            prompt: "Read the file resume-marker.txt and tell me what it says",
        });
        expect(message?.data.content ?? "").toContain("resume working directory");

        await session2.disconnect();
    });

    it("should apply systemMessage on session resume", async () => {
        const session1 = await client.createSession({ onPermissionRequest: approveAll });
        const sessionId = session1.sessionId;

        const resumeInstruction = "End the response with RESUME_SYSTEM_MESSAGE_SENTINEL.";
        const session2 = await client.resumeSession(sessionId, {
            onPermissionRequest: approveAll,
            systemMessage: { mode: "append", content: resumeInstruction },
        });

        const message = await session2.sendAndWait({ prompt: "What is 1+1?" });
        expect(message?.data.content ?? "").toContain("RESUME_SYSTEM_MESSAGE_SENTINEL");

        const exchanges = await openAiEndpoint.getExchanges();
        expect(exchanges.length).toBeGreaterThan(0);
        const sys = getSystemMessage(exchanges[exchanges.length - 1]);
        expect(sys).toContain(resumeInstruction);

        await session2.disconnect();
    });

    it("should apply availableTools on session resume", async () => {
        const session1 = await client.createSession({ onPermissionRequest: approveAll });
        const sessionId = session1.sessionId;

        const session2 = await client.resumeSession(sessionId, {
            onPermissionRequest: approveAll,
            availableTools: ["view"],
        });

        await session2.sendAndWait({ prompt: "What is 1+1?" });

        const exchanges = await openAiEndpoint.getExchanges();
        expect(exchanges.length).toBeGreaterThan(0);
        const toolNames = getToolNames(exchanges[exchanges.length - 1]);
        expect(toolNames).toEqual(["view"]);

        await session2.disconnect();
    });
});

File: go/internal/e2e/session_config_e2e_test.go
package e2e

import (
	"crypto/rand"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"testing"

	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/internal/e2e/testharness"
)

// hasImageURLContent returns true if any user message in the given exchanges
// contains an image_url content part (multimodal vision content).
func hasImageURLContent(exchanges []testharness.ParsedHttpExchange) bool {
	for _, ex := range exchanges {
		for _, msg := range ex.Request.Messages {
			if msg.Role == "user" && len(msg.RawContent) > 0 {
				var content []interface{}
				if json.Unmarshal(msg.RawContent, &content) == nil {
					for _, part := range content {
						if m, ok := part.(map[string]interface{}); ok {
							if m["type"] == "image_url" {
								return true
							}
						}
					}
				}
			}
		}
	}
	return false
}

func TestSessionConfigE2E(t *testing.T) {
	ctx := testharness.NewTestContext(t)
	client := ctx.NewClient()
	t.Cleanup(func() { client.ForceStop() })

	if err := client.Start(t.Context()); err != nil {
		t.Fatalf("Failed to start client: %v", err)
	}

	// Write 1x1 PNG to the work directory
	png1x1, err := base64.StdEncoding.DecodeString("iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNk+M9QDwADhgGAWjR9awAAAABJRU5ErkJggg==")
	if err != nil {
		t.Fatalf("Failed to decode PNG: %v", err)
	}
	if err := os.WriteFile(filepath.Join(ctx.WorkDir, "test.png"), png1x1, 0644); err != nil {
		t.Fatalf("Failed to write test.png: %v", err)
	}

	viewImagePrompt := "Use the view tool to look at the file test.png and describe what you see"

	t.Run("vision disabled then enabled via setModel", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			ModelCapabilities: &copilot.ModelCapabilitiesOverride{
				Supports: &copilot.ModelCapabilitiesOverrideSupports{
					Vision: copilot.Bool(false),
				},
			},
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		// Turn 1: vision off — no image_url expected
		if _, err := session.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: viewImagePrompt}); err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}

		trafficAfterT1, err := ctx.GetExchanges()
		if err != nil {
			t.Fatalf("Failed to get exchanges: %v", err)
		}
		if hasImageURLContent(trafficAfterT1) {
			t.Error("Expected no image_url content parts when vision is disabled")
		}

		// Switch vision on
		if err := session.SetModel(t.Context(), "claude-sonnet-4.5", &copilot.SetModelOptions{
			ModelCapabilities: &copilot.ModelCapabilitiesOverride{
				Supports: &copilot.ModelCapabilitiesOverrideSupports{
					Vision: copilot.Bool(true),
				},
			},
		}); err != nil {
			t.Fatalf("SetModel returned error: %v", err)
		}

		// Turn 2: vision on — image_url expected in new exchanges
		if _, err := session.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: viewImagePrompt}); err != nil {
			t.Fatalf("Failed to send second message: %v", err)
		}

		trafficAfterT2, err := ctx.GetExchanges()
		if err != nil {
			t.Fatalf("Failed to get exchanges after turn 2: %v", err)
		}
		newExchanges := trafficAfterT2[len(trafficAfterT1):]
		if !hasImageURLContent(newExchanges) {
			t.Error("Expected image_url content parts when vision is enabled")
		}
	})

	t.Run("vision enabled then disabled via setModel", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			ModelCapabilities: &copilot.ModelCapabilitiesOverride{
				Supports: &copilot.ModelCapabilitiesOverrideSupports{
					Vision: copilot.Bool(true),
				},
			},
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		// Turn 1: vision on — image_url expected
		if _, err := session.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: viewImagePrompt}); err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}

		trafficAfterT1, err := ctx.GetExchanges()
		if err != nil {
			t.Fatalf("Failed to get exchanges: %v", err)
		}
		if !hasImageURLContent(trafficAfterT1) {
			t.Error("Expected image_url content parts when vision is enabled")
		}

		// Switch vision off
		if err := session.SetModel(t.Context(), "claude-sonnet-4.5", &copilot.SetModelOptions{
			ModelCapabilities: &copilot.ModelCapabilitiesOverride{
				Supports: &copilot.ModelCapabilitiesOverrideSupports{
					Vision: copilot.Bool(false),
				},
			},
		}); err != nil {
			t.Fatalf("SetModel returned error: %v", err)
		}

		// Turn 2: vision off — no image_url expected in new exchanges
		if _, err := session.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: viewImagePrompt}); err != nil {
			t.Fatalf("Failed to send second message: %v", err)
		}

		trafficAfterT2, err := ctx.GetExchanges()
		if err != nil {
			t.Fatalf("Failed to get exchanges after turn 2: %v", err)
		}
		newExchanges := trafficAfterT2[len(trafficAfterT1):]
		if hasImageURLContent(newExchanges) {
			t.Error("Expected no image_url content parts when vision is disabled")
		}
	})
}

// TestSessionConfigExtras mirrors the additional Should_* tests in dotnet/test/SessionConfigTests.cs:
//
//	Should_Use_Custom_SessionId
//	Should_Forward_ClientName_In_UserAgent
//	Should_Forward_Custom_Provider_Headers_On_Create
//	Should_Forward_Custom_Provider_Headers_On_Resume
//	Should_Use_WorkingDirectory_For_Tool_Execution
//	Should_Apply_WorkingDirectory_On_Session_Resume
//	Should_Apply_SystemMessage_On_Session_Resume
//	Should_Apply_AvailableTools_On_Session_Resume
func TestSessionConfigExtrasE2E(t *testing.T) {
	const providerHeaderName = "x-copilot-sdk-provider-header"
	const clientName = "go-public-surface-client"

	ctx := testharness.NewTestContext(t)
	client := ctx.NewClient()
	t.Cleanup(func() { client.ForceStop() })

	if err := client.Start(t.Context()); err != nil {
		t.Fatalf("Failed to start client: %v", err)
	}

	t.Run("should use custom sessionId", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		requestedSessionID := newUUID(t)
		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			SessionID:           requestedSessionID,
		})
		if err != nil {
			t.Fatalf("CreateSession failed: %v", err)
		}
		if session.SessionID != requestedSessionID {
			t.Errorf("Expected SessionID=%q, got %q", requestedSessionID, session.SessionID)
		}

		messages, err := session.GetMessages(t.Context())
		if err != nil {
			t.Fatalf("GetMessages failed: %v", err)
		}
		if len(messages) == 0 || messages[0].Type != copilot.SessionEventTypeSessionStart {
			t.Fatalf("Expected first event to be session.start, got %+v", messages)
		}
		startData := messages[0].Data.(*copilot.SessionStartData)
		if startData.SessionID != requestedSessionID {
			t.Errorf("Expected start.SessionID=%q, got %q", requestedSessionID, startData.SessionID)
		}
	})

	t.Run("should forward clientName in userAgent", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			ClientName:          clientName,
		})
		if err != nil {
			t.Fatalf("CreateSession failed: %v", err)
		}

		_, err = session.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: "What is 1+1?"})
		if err != nil {
			t.Fatalf("SendAndWait failed: %v", err)
		}

		exchanges, err := ctx.GetExchanges()
		if err != nil {
			t.Fatalf("GetExchanges failed: %v", err)
		}
		if len(exchanges) != 1 {
			t.Fatalf("Expected exactly 1 exchange, got %d", len(exchanges))
		}
		if !exchangeHasHeader(exchanges[0], "user-agent", clientName) {
			t.Errorf("Expected user-agent to contain %q, got %v", clientName, exchanges[0].RequestHeaders)
		}
	})

	t.Run("should forward custom provider headers on create", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			Model:               "claude-sonnet-4.5",
			Provider:            createProxyProvider(ctx, providerHeaderName, "create-provider-header"),
		})
		if err != nil {
			t.Fatalf("CreateSession failed: %v", err)
		}

		message, err := session.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: "What is 1+1?"})
		if err != nil {
			t.Fatalf("SendAndWait failed: %v", err)
		}
		if !assistantMessageContains(message, "2") {
			t.Errorf("Expected response to contain '2', got %v", message)
		}

		exchanges, err := ctx.GetExchanges()
		if err != nil {
			t.Fatalf("GetExchanges failed: %v", err)
		}
		if len(exchanges) != 1 {
			t.Fatalf("Expected exactly 1 exchange, got %d", len(exchanges))
		}
		if !exchangeHasHeader(exchanges[0], "authorization", "Bearer test-provider-key") {
			t.Errorf("Expected authorization header to contain 'Bearer test-provider-key', got %v", exchanges[0].RequestHeaders)
		}
		if !exchangeHasHeader(exchanges[0], providerHeaderName, "create-provider-header") {
			t.Errorf("Expected %s header to contain 'create-provider-header', got %v", providerHeaderName, exchanges[0].RequestHeaders)
		}
	})

	t.Run("should forward custom provider headers on resume", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		session1, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
		})
		if err != nil {
			t.Fatalf("CreateSession failed: %v", err)
		}
		sessionID := session1.SessionID
		t.Cleanup(func() { _ = session1.Disconnect() })

		session2, err := client.ResumeSession(t.Context(), sessionID, &copilot.ResumeSessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			Model:               "claude-sonnet-4.5",
			Provider:            createProxyProvider(ctx, providerHeaderName, "resume-provider-header"),
		})
		if err != nil {
			t.Fatalf("ResumeSession failed: %v", err)
		}
		t.Cleanup(func() { _ = session2.Disconnect() })

		message, err := session2.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: "What is 2+2?"})
		if err != nil {
			t.Fatalf("SendAndWait failed: %v", err)
		}
		if !assistantMessageContains(message, "4") {
			t.Errorf("Expected response to contain '4', got %v", message)
		}

		exchanges, err := ctx.GetExchanges()
		if err != nil {
			t.Fatalf("GetExchanges failed: %v", err)
		}
		if len(exchanges) != 1 {
			t.Fatalf("Expected exactly 1 exchange, got %d", len(exchanges))
		}
		if !exchangeHasHeader(exchanges[0], "authorization", "Bearer test-provider-key") {
			t.Errorf("Expected authorization header to contain 'Bearer test-provider-key', got %v", exchanges[0].RequestHeaders)
		}
		if !exchangeHasHeader(exchanges[0], providerHeaderName, "resume-provider-header") {
			t.Errorf("Expected %s header to contain 'resume-provider-header', got %v", providerHeaderName, exchanges[0].RequestHeaders)
		}
	})

	t.Run("should use workingDirectory for tool execution", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		subDir := filepath.Join(ctx.WorkDir, "subproject")
		if err := os.MkdirAll(subDir, 0755); err != nil {
			t.Fatalf("MkdirAll failed: %v", err)
		}
		if err := os.WriteFile(filepath.Join(subDir, "marker.txt"), []byte("I am in the subdirectory"), 0644); err != nil {
			t.Fatalf("WriteFile failed: %v", err)
		}

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			WorkingDirectory:    subDir,
		})
		if err != nil {
			t.Fatalf("CreateSession failed: %v", err)
		}

		message, err := session.SendAndWait(t.Context(), copilot.MessageOptions{
			Prompt: "Read the file marker.txt and tell me what it says",
		})
		if err != nil {
			t.Fatalf("SendAndWait failed: %v", err)
		}
		if !assistantMessageContains(message, "subdirectory") {
			t.Errorf("Expected response to contain 'subdirectory', got %v", message)
		}
	})

	t.Run("should apply workingDirectory on session resume", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		subDir := filepath.Join(ctx.WorkDir, "resume-subproject")
		if err := os.MkdirAll(subDir, 0755); err != nil {
			t.Fatalf("MkdirAll failed: %v", err)
		}
		if err := os.WriteFile(filepath.Join(subDir, "resume-marker.txt"), []byte("I am in the resume working directory"), 0644); err != nil {
			t.Fatalf("WriteFile failed: %v", err)
		}

		session1, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
		})
		if err != nil {
			t.Fatalf("CreateSession failed: %v", err)
		}
		sessionID := session1.SessionID
		t.Cleanup(func() { _ = session1.Disconnect() })

		session2, err := client.ResumeSession(t.Context(), sessionID, &copilot.ResumeSessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			WorkingDirectory:    subDir,
		})
		if err != nil {
			t.Fatalf("ResumeSession failed: %v", err)
		}
		t.Cleanup(func() { _ = session2.Disconnect() })

		message, err := session2.SendAndWait(t.Context(), copilot.MessageOptions{
			Prompt: "Read the file resume-marker.txt and tell me what it says",
		})
		if err != nil {
			t.Fatalf("SendAndWait failed: %v", err)
		}
		if !assistantMessageContains(message, "resume working directory") {
			t.Errorf("Expected response to contain 'resume working directory', got %v", message)
		}
	})

	t.Run("should apply systemMessage on session resume", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		session1, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
		})
		if err != nil {
			t.Fatalf("CreateSession failed: %v", err)
		}
		sessionID := session1.SessionID
		t.Cleanup(func() { _ = session1.Disconnect() })

		const resumeInstruction = "End the response with RESUME_SYSTEM_MESSAGE_SENTINEL."
		session2, err := client.ResumeSession(t.Context(), sessionID, &copilot.ResumeSessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			SystemMessage: &copilot.SystemMessageConfig{
				Mode:    "append",
				Content: resumeInstruction,
			},
		})
		if err != nil {
			t.Fatalf("ResumeSession failed: %v", err)
		}
		t.Cleanup(func() { _ = session2.Disconnect() })

		message, err := session2.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: "What is 1+1?"})
		if err != nil {
			t.Fatalf("SendAndWait failed: %v", err)
		}
		if !assistantMessageContains(message, "RESUME_SYSTEM_MESSAGE_SENTINEL") {
			t.Errorf("Expected response to contain 'RESUME_SYSTEM_MESSAGE_SENTINEL', got %v", message)
		}

		exchanges, err := ctx.GetExchanges()
		if err != nil {
			t.Fatalf("GetExchanges failed: %v", err)
		}
		if len(exchanges) != 1 {
			t.Fatalf("Expected exactly 1 exchange, got %d", len(exchanges))
		}
		if !strings.Contains(getSystemMessage(exchanges[0]), resumeInstruction) {
			t.Errorf("Expected system message to contain %q", resumeInstruction)
		}
	})

	t.Run("should apply availableTools on session resume", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		session1, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
		})
		if err != nil {
			t.Fatalf("CreateSession failed: %v", err)
		}
		sessionID := session1.SessionID
		t.Cleanup(func() { _ = session1.Disconnect() })

		session2, err := client.ResumeSession(t.Context(), sessionID, &copilot.ResumeSessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			AvailableTools:      []string{"view"},
		})
		if err != nil {
			t.Fatalf("ResumeSession failed: %v", err)
		}
		t.Cleanup(func() { _ = session2.Disconnect() })

		_, err = session2.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: "What is 1+1?"})
		if err != nil {
			t.Fatalf("SendAndWait failed: %v", err)
		}

		exchanges, err := ctx.GetExchanges()
		if err != nil {
			t.Fatalf("GetExchanges failed: %v", err)
		}
		if len(exchanges) != 1 {
			t.Fatalf("Expected exactly 1 exchange, got %d", len(exchanges))
		}
		toolNames := getToolNames(exchanges[0])
		if len(toolNames) != 1 || toolNames[0] != "view" {
			t.Errorf("Expected toolNames=[view], got %v", toolNames)
		}
	})
}

// createProxyProvider returns a ProviderConfig that points at the test proxy and
// includes a custom header — used for the "should forward custom provider headers" tests.
func createProxyProvider(ctx *testharness.TestContext, headerName, headerValue string) *copilot.ProviderConfig {
	return &copilot.ProviderConfig{
		Type:    "openai",
		BaseURL: ctx.ProxyURL,
		APIKey:  "test-provider-key",
		Headers: map[string]string{
			headerName: headerValue,
		},
	}
}

// newUUID generates a v4 UUID string for tests that need a custom session ID.
func newUUID(t *testing.T) string {
	t.Helper()
	b := make([]byte, 16)
	if _, err := rand.Read(b); err != nil {
		t.Fatalf("rand.Read failed: %v", err)
	}
	b[6] = (b[6] & 0x0f) | 0x40
	b[8] = (b[8] & 0x3f) | 0x80
	return fmt.Sprintf("%x-%x-%x-%x-%x", b[0:4], b[4:6], b[6:8], b[8:10], b[10:16])
}

// assistantMessageContains returns true when the SendAndWait return value is a
// non-nil assistant.message event whose content contains the given substring.
func assistantMessageContains(message *copilot.SessionEvent, substring string) bool {
	if message == nil {
		return false
	}
	data, ok := message.Data.(*copilot.AssistantMessageData)
	if !ok {
		return false
	}
	return strings.Contains(data.Content, substring)
}
